client-tez/src/main/java/org/apache/tez/runtime/library/input/RssOrderedGroupedKVInput.java

```diff
@@ -0,0 +1,428 @@
+            TaskCounter.REDUCE_INPUT_RECORDS);
+    this.shuffledInputs = getContext().getCounters().findCounter(
+            TaskCounter.NUM_SHUFFLED_INPUTS);
+    this.conf.setStrings(TezRuntimeFrameworkConfigs.LOCAL_DIRS, getContext().getWorkDirs());
+
+    return Collections.emptyList();
+  }
+
+  @Override
+  public synchronized void start() throws IOException {
+    if (!isStarted.get()) {
+      memoryUpdateCallbackHandler.validateUpdateReceived();
+      // Start the shuffle - copy and merge
+      shuffle = createRssShuffle();
+      shuffle.run();
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Initialized the handlers in shuffle..Safe to start processing..");
+      }
+      List<Event> pending = new LinkedList<>();
+      pendingEvents.drainTo(pending);
+      if (pending.size() > 0) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("NoAutoStart delay in processing first event: "
+                  + (System.currentTimeMillis() - firstEventReceivedTime));
+        }
+        shuffle.handleEvents(pending);
+      }
+      isStarted.set(true);
+    }
+  }
+
+  @VisibleForTesting
+  RssShuffle createRssShuffle() throws IOException {
+    long initialMemoryAvailable = conf.getLong(RssTezConfig.RSS_REDUCE_INITIAL_MEMORY, 200 * 1024 * 1024L);
+    LOG.info("RssOrderedGroupedKVInput, initialMemoryAvailable from config:{}, final initialMemoryAvailable:{}",
+        conf.get(RssTezConfig.RSS_REDUCE_INITIAL_MEMORY), initialMemoryAvailable);
+    return new RssShuffle(getContext(), conf, getNumPhysicalInputs(), memoryUpdateCallbackHandler.getMemoryAssigned());
+  }
+
+  /**
+   * Check if the input is ready for consumption
+   *
+   * @return true if the input is ready for consumption, or if an error occurred
+   *         processing fetching the input. false if the shuffle and merge are
+   *         still in progress
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  public synchronized boolean isInputReady() throws IOException, InterruptedException, TezException {
+    Preconditions.checkState(isStarted.get(), "Must start input before invoking this method");
+    if (getNumPhysicalInputs() == 0) {
+      return true;
+    }
+    return shuffle.isInputReady();
+  }
+
+  /**
+   * Waits for the input to become ready for consumption
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public void waitForInputReady() throws IOException, InterruptedException, TezException {
+    // Cannot synchronize entire method since this is called form user code and can block.
+    RssShuffle localShuffleCopy = null;
+    synchronized (this) {
+      Preconditions.checkState(isStarted.get(), "Must start input before invoking this method");
+      if (getNumPhysicalInputs() == 0) {
+        return;
+      }
+      localShuffleCopy = shuffle;
+    }
+
+    TezRawKeyValueIterator localRawIter = localShuffleCopy.waitForInput();
+    synchronized (this) {
```

Comment:
Will the class be accessed by multiple threads?

Reply:
Yes，It will create by LogicalIOProcessorRuntimeTask in  InitializeInputCallable Thread:
(When the vertex has multi up-vertex, it will create multi Input. That is to say, one vertex is one input)

https://github.com/apache/tez/blob/ccad71dabceb850b58ff38e77fd206e68f62ce4b/tez-runtime-internals/src/main/java/org/apache/tez/runtime/LogicalIOProcessorRuntimeTask.java#L448